### PR TITLE
[Model Runner V2] Fix rejection sampling acceptance rate gap vs MRV1

### PR DESCRIPTION
**Pull request by TheEpicDolphin**
# Context
Currently, MRV1 does not use the true draft logits for rejection sampling. Instead, it [samples draft tokens via argmax](https://github.com/vllm-project/vllm/blob/22fa63cfe849c38bbdb590163043230a1a1a6148/vllm/v1/spec_decode/eagle.py#L398C9-L398C23), essentially treating the draft probability distribution, q(x), as one-hot. This changes the standard rejection sampling ratio probability test from u < p(x) / q(x) to u < p(x), as shown [here](https://github.com/vllm-project/vllm/blob/22fa63cfe849c38bbdb590163043230a1a1a6148/vllm/v1/sample/rejection_sampler.py#L738). If the token is rejected, it samples from the residual correction which is just p(x) with the draft token's probability zero-ed out, respecting the target distribution.

MRV2 currently uses strict acceptance by default, which only accepts a draft token if it exactly matches the target sampled token. This is suboptimal for rejection sampling acceptance rates. MRV2 also has a WIP "probabilistic" rejection sampling feature, but it uses the true, full draft probabilities for potentially higher acceptance rates, at the cost of more memory usage.

# This PR
Achieve parity with MRV1's rejection sampling by enabling the "middle-ground" rejection sampling in MRV2 which greedily samples draft tokens and treats the draft distribution as one-hot. This yields acceptance rates on par with MRV1, while not storing the full draft logits.

I added a separate parameter called "draft_sample_method". When set to `greedy`, argmax is used for the draft tokens, like in MRV1. When set to `gumbel`, gumbel sampling is used and we do rejection sampling with the full draft probabilities. By default, this parameter is set to `greedy`.

# Benchmarks
**Command**
```
VLLM_USE_V2_MODEL_RUNNER=0 vllm serve Qwen/Qwen3-8B \
  --no-enable-prefix-caching \
  --tensor-parallel-size 1 \
  --data-parallel-size 1 \
  --speculative-config '{"method": "eagle3", "model": "RedHatAI/Qwen3-8B-speculator.eagle3", "num_speculative_tokens": 7}'
```

## mt-bench, temperature = 0.0
Metric | MRV1 | MRV2 Before | MRV2 After
-- | -- | -- | --
Successful requests | 1000 | 1000 | 1000
Failed requests | 0 | 0 | 0
Maximum request concurrency | 64 | 64 | 64
Benchmark duration (s) | 18.80 | 15.92 | 15.81
Total input tokens | 67464 | 67464 | 67464
Total generated tokens | 128000 | 128000 | 128000
Request throughput (req/s) | 53.18 | 62.81 | 63.26
Output token throughput (tok/s) | 6807.12 | 8039.99 | 8097.38
Peak output token throughput (tok/s) | 3458.00 | 3531.00 | 3582.00
Peak concurrent requests | 131.00 | 132.00 | 133.00
Total token throughput (tok/s) | 10394.90 | 12277.57 | 12365.21
TTFT |   |   |  
Mean TTFT (ms) | 84.94 | 70.67 | 67.16
Median TTFT (ms) | 58.32 | 57.53 | 57.29
P99 TTFT (ms) | 729.46 | 220.55 | 226.36
P90 TTFT (ms) | 76.07 | 98.83 | 70.28
TPOT |   |   |  
Mean TPOT (ms) | 7.55 | 7.25 | 7.26
Median TPOT (ms) | 7.36 | 7.17 | 7.19
P99 TPOT (ms) | 11.78 | 9.94 | 10.04
P90 TPOT (ms) | 8.99 | 8.58 | 8.52
ITL |   |   |  
Mean ITL (ms) | 18.35 | 18.06 | 17.66
Median ITL (ms) | 18.57 | 17.31 | 17.85
P99 ITL (ms) | 29.69 | 33.03 | 28.00
P90 ITL (ms) | 21.69 | 22.04 | 21.28
E2EL |   |   |  
Mean E2EL (ms) | 1043.23 | 991.74 | 989.02
Median E2EL (ms) | 1001.96 | 980.03 | 976.66
P99 E2EL (ms) | 1758.04 | 1467.56 | 1461.79
P90 E2EL (ms) | 1245.77 | 1170.54 | 1151.91
Speculative Decoding |   |   |  
Acceptance rate (%) | 21.52 | 21.51 | 21.53
Acceptance length | 2.51 | 2.51 | 2.51
Drafts | 51217 | 51208 | 51202
Draft tokens | 358519 | 358456 | 358414
Accepted tokens | 77153 | 77110 | 77154
Position 0 acceptance (%) | 66.67 | 66.62 | 66.65
Position 1 acceptance (%) | 40.48 | 40.64 | 40.68
Position 2 acceptance (%) | 20.75 | 20.64 | 20.67
Position 3 acceptance (%) | 10.74 | 10.64 | 10.64
Position 4 acceptance (%) | 6.09 | 6.09 | 6.10
Position 5 acceptance (%) | 3.63 | 3.66 | 3.65
Position 6 acceptance (%) | 2.27 | 2.30 | 2.30

## mt-bench, temperature = 1.0
Metric | MRV1 | MRV2 Before | MRV2 After
-- | -- | -- | --
Successful requests | 1000 | 1000 | 1000
Failed requests | 0 | 0 | 0
Maximum request concurrency | 64 | 64 | 64
Benchmark duration (s) | 18.45 | 19.46 | 16.98
Total input tokens | 67464 | 67464 | 67464
Total generated tokens | 128000 | 128000 | 128000
Request throughput (req/s) | 54.19 | 51.38 | 58.90
Output token throughput (tok/s) | 6936.04 | 6576.37 | 7538.83
Peak output token throughput (tok/s) | 3177.00 | 3152.00 | 3454.00
Peak concurrent requests | 126.00 | 129.00 | 129.00
Total token throughput (tok/s) | 10591.77 | 10042.53 | 11512.27
TTFT |   |   |  
Mean TTFT (ms) | 100.60 | 70.85 | 69.23
Median TTFT (ms) | 62.71 | 62.69 | 59.63
P99 TTFT (ms) | 728.42 | 202.44 | 264.41
P90 TTFT (ms) | 78.58 | 79.71 | 77.62
TPOT |   |   |  
Mean TPOT (ms) | 8.33 | 9.07 | 7.84
Median TPOT (ms) | 8.34 | 9.13 | 7.91
P99 TPOT (ms) | 11.42 | 11.97 | 10.85
P90 TPOT (ms) | 9.91 | 10.61 | 9.19
ITL |   |   |  
Mean ITL (ms) | 19.92 | 19.90 | 18.72
Median ITL (ms) | 19.64 | 19.55 | 18.48
P99 ITL (ms) | 34.30 | 35.86 | 33.44
P90 ITL (ms) | 23.63 | 23.40 | 22.31
E2EL |   |   |  
Mean E2EL (ms) | 1158.92 | 1222.96 | 1065.17
Median E2EL (ms) | 1135.27 | 1228.25 | 1074.25
P99 E2EL (ms) | 1866.05 | 1587.92 | 1455.58
P90 E2EL (ms) | 1397.99 | 1425.59 | 1242.41
Speculative Decoding |   |   |  
Acceptance rate (%) | 20.82 | 17.55 | 20.63
Acceptance length | 2.46 | 2.23 | 2.44
Drafts | 52129 | 57417 | 52402
Draft tokens | 364903 | 401919 | 366814
Accepted tokens | 75975 | 70535 | 75670
Position 0 acceptance (%) | 64.71 | 59.84 | 64.53
Position 1 acceptance (%) | 39.10 | 33.09 | 39.10
Position 2 acceptance (%) | 19.98 | 15.05 | 19.60
Position 3 acceptance (%) | 10.40 | 7.41 | 9.97
Position 4 acceptance (%) | 5.83 | 3.88 | 5.66
Position 5 acceptance (%) | 3.52 | 2.22 | 3.42
Position 6 acceptance (%) | 2.19 | 1.36 | 2.14

# Future Work
synthetic rejection sampling needs to be modified so that it takes into account the performance of standard rejection sampling (with or without draft logits).

### CODE REVIEW

**Review by benchislett — COMMENTED** [review]
Need to double-check the math but at a high level it looks good
